File: src/org/genyris/web/HTTPgetFunction.java
// Copyright 2009 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.web;

import java.io.IOException;

import org.apache.http.HttpVersion;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.CloseableHttpClient;

import org.genyris.core.*;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.Closure;
import org.genyris.interp.Environment;
import org.genyris.interp.Interpreter;

public class HTTPgetFunction extends HTTPclientFunction {

    public HTTPgetFunction(Interpreter interp) {
        super(interp, Constants.WEB + "get", true);
    }


    public Exp bindAndExecute(Closure proc, Exp[] arguments,
            Environment envForBindOperations) throws GenyrisException {

        String URI = getArg(arguments, 0, StrinG.class, true).toString();
        Exp headers = getArg(arguments, 1, Exp.class);
        Exp protocol = getArg(arguments, 2, Exp.class);
        Exp options = getArg(arguments, 3, Symbol.class);

        CloseableHttpClient httpclient = getCloseableHttpClient(options);
        HttpVersion httpVersion = parseProtocol(protocol);

        try {

            HttpGet httpGet = new HttpGet(URI);
            httpGet.setProtocolVersion(httpVersion);

            addHeadersToRequest(headers, charset, httpGet);

            CloseableHttpResponse response = httpclient.execute(httpGet);

            return processResponse(URI, response);
        } catch (IOException e) {
            throw new GenyrisException(e.getMessage());
        } catch (java.lang.RuntimeException e) {
            throw new GenyrisException(e.toString());
        }
        finally {
            try {
                httpclient.close();
            } catch (IOException ignored) {
                ignored.printStackTrace();
            }
        }

    }

}

File: src/org/genyris/web/HTTPgetFunction.kt
// Copyright 2009 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.web

import org.apache.http.client.methods.HttpGet
import org.apache.http.impl.client.CloseableHttpClient
import org.genyris.core.Exp
import org.genyris.core.StrinG
import org.genyris.core.Symbol
import org.genyris.exception.GenyrisException
import org.genyris.interp.Closure
import org.genyris.interp.Environment
import org.genyris.interp.Interpreter
import java.io.IOException

class HTTPgetFunction(interp: Interpreter?) : HTTPclientFunction(interp, "get", true) {
    @kotlin.Throws(GenyrisException::class)
    override fun bindAndExecute(
        proc: Closure?, arguments: Array<Exp?>,
        envForBindOperations: Environment?
    ): Exp? {
        val URI: String? = getArg(arguments, 0, StrinG::class.java, true).toString()
        val headers = getArg(arguments, 1, Exp::class.java)
        val protocol = getArg(arguments, 2, Exp::class.java)
        val options = getArg(arguments, 3, Symbol::class.java)

        val httpclient: CloseableHttpClient = HTTPclientFunction.Companion.getCloseableHttpClient(options)
        val httpVersion = parseProtocol(protocol)

        try {
            val httpGet = HttpGet(URI)
            httpGet.setProtocolVersion(httpVersion)

            addHeadersToRequest(headers, charset, httpGet)

            val response = httpclient.execute(httpGet)

            return processResponse(URI, response)
        } catch (e: IOException) {
            throw GenyrisException(e.getMessage())
        } catch (e: RuntimeException) {
            throw GenyrisException(e.toString())
        } finally {
            try {
                httpclient.close()
            } catch (ignored: IOException) {
                ignored.printStackTrace()
            }
        }
    }
}
